src/vmspawn/vmspawn.c

```diff
@@ -208,11 +261,189 @@ static int parse_argv(int argc, char *argv[]) {
+                r = sd_notify(false, "READY=1\n");
+                if (r < 0)
+                        log_warning_errno(r, "Failed to send readiness notification, ignoring: %m");
+        }
+
+        p = strv_find_startswith(tags, "STATUS=");
+        if (p)
+                (void) sd_notifyf(false, "STATUS=VM running: %s", p);
+
+        p = strv_find_startswith(tags, "EXIT_STATUS=");
+        if (p) {
+                r = safe_atoi(p, exit_status);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to parse exit status from %s, ignoring: %m", p);
+        }
+
+        /* we will only receive one message from each connection so disable this source once one is received */
+        source = sd_event_source_disable_unref(source);
+
+        return 0;
+}
+
+static int vmspawn_dispatch_vsock_connections(sd_event_source *source, int fd, uint32_t revents, void *userdata) {
+        int r;
+        sd_event *event;
+        _cleanup_close_ int conn_fd = -EBADF;
+
+        assert(userdata);
+
+        if (revents != EPOLLIN) {
+                log_warning("Got unexpected poll event for vsock fd.");
+                return 0;
+        }
+
+        conn_fd = accept4(fd, NULL, NULL, SOCK_CLOEXEC|SOCK_NONBLOCK);
+        if (conn_fd < 0) {
+                log_warning_errno(errno, "Failed to accept connection from vsock fd (%m), ignoring...");
+                return 0;
+        }
+
+        event = sd_event_source_get_event(source);
+        if (!event)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOENT), "Failed to retrieve event from event source, exiting task");
+
+        /* add a new floating task to read from the connection */
+        r = sd_event_add_io(event, NULL, conn_fd, revents, vmspawn_dispatch_notify_fd, userdata);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate notify connection event source: %m");
+
+        /* conn_fd is now owned by the event loop so don't clean it up */
+        TAKE_FD(conn_fd);
+
+        return 0;
+}
+
+static int setup_notify_parent(sd_event *event, int fd, int *exit_status, sd_event_source **notify_event_source) {
+        int r;
+
+        r = sd_event_add_io(event, notify_event_source, fd, EPOLLIN, vmspawn_dispatch_vsock_connections, exit_status);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate notify socket event source: %m");
+
+        (void) sd_event_source_set_description(*notify_event_source, "vmspawn-notify-sock");
+
+        return 0;
+}
+
+static int on_orderly_shutdown(sd_event_source *s, const struct signalfd_siginfo *si, void *userdata) {
+        pid_t pid;
+
+        pid = PTR_TO_PID(userdata);
+        if (pid > 0) {
+                /* TODO: actually talk to qemu and ask the guest to shutdown here */
+                if (kill(pid, SIGKILL) >= 0) {
+                        log_info("Trying to halt qemu. Send SIGTERM again to trigger immediate termination.");
```

Comment:
code does SIGKILL, log msg says SIGTERM

Reply:
The sigkill is killing QEMU, the SIGTERM is to stop vmspawn from hanging. I'll change the message to make it clearer that we are just killing vmspawn from a second SIGTERM.

Reply:
Fixed in ef3bb39e80.